core/commands/storage.go

```diff
@@ -544,31 +676,47 @@ func prepareSignedContractsForShard(param *paramsForPrepareContractsForShard, ca
 	return nil
 }
 
-func prepareSignedContractsForShardOffSign(param *paramsForPrepareContractsForShard,
-	candidateHost *storage.HostNode, initialCall bool) error {
-	shard := param.shard
-	shard.CandidateHost = candidateHost
+func buildContractsForShard(param *paramsForPrepareContractsForShard,
+	candidateHost *storage.HostNode) (*escrowpb.EscrowContract, *guardpb.ContractMeta, error) {
 	ss := param.ss
+	shard := param.shard
+	shardIndex := shard.ShardIndex
 
-	err := prepareSignedContractsForEscrowOffSign(param, !initialCall)
+	shard.SetPrice(candidateHost.Price)
+	cfg, err := param.n.Repo.Config()
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 
-	err = prepareSignedGuardContractForShardOffSign(ss, shard, param.n, !initialCall)
+	// init escrow/guard Contracts
+	_, hostPid, err := ParsePeerParam(candidateHost.Identity)
 	if err != nil {
-		return err
+		return nil, nil, err
+	}
+	var offSignPid peer.ID
+	if ss.IsOffSignRunmode() {
+		if ss.OfflineCB == nil {
+			return nil, nil, errors.New("unexpected nil valure for ss.OfflineCB")
+		}
+		offSignPid = ss.OfflineCB.OfflinePeerID
+	} else {
+		offSignPid = ""
+	}
+	escrowContract, err := escrow.NewContract(cfg, shard.ContractID, param.n, hostPid,
+		shard.TotalPay, param.customizedSchedule, param.period, offSignPid)
+	if err != nil {
+		return nil, nil, fmt.Errorf("create escrow contract failed: [%v] ", err)
 	}
 
-	err = buildSignedGuardContractForShardOffSign(ss, shard, param.n, param.runMode,
-		ss.OfflineCB.OfflinePeerID.Pretty(), !initialCall)
+	shard.UpdateShard(hostPid)
+	guardContractMeta, err := guard.NewContract(ss, cfg, param.shardKey, int32(shardIndex), param.renterPid)
 	if err != nil {
-		return err
+		return nil, nil, fmt.Errorf("fail to new contract meta: [%v] ", err)
 	}
-	return nil
+	return escrowContract, guardContractMeta, nil
 }
 
-func retryMonitor(ctx context.Context, api coreiface.CoreAPI, ss *storage.FileContracts, n *core.IpfsNode,
+func retryMonitor(monitorCtx context.Context, api coreiface.CoreAPI, ss *storage.FileContracts, n *core.IpfsNode,
```

Comment:
You can remove `monitorCtx` and use the central `ss.RetryMonitorCtx` so it's easier to do timeout later.